File: ruby/mercuryledger/tests/services/audit_service_test.rb
# frozen_string_literal: true

require_relative '../test_helper'
require_relative '../../services/audit/service'

class AuditServiceTest < Minitest::Test
  def test_validate_audit_entry_accepts_valid
    entry = MercuryLedger::Services::Audit::AuditEntry.new(
      service: 'gateway', action: 'deploy', severity: 3, timestamp: Time.now.to_i, details: 'ok'
    )
    assert MercuryLedger::Services::Audit.validate_audit_entry(entry)
  end

  def test_summarize_trail_counts_entries
    entries = [
      MercuryLedger::Services::Audit::AuditEntry.new(service: 'gw', action: 'a', severity: 2, timestamp: 1),
      MercuryLedger::Services::Audit::AuditEntry.new(service: 'rt', action: 'b', severity: 4, timestamp: 2)
    ]
    summary = MercuryLedger::Services::Audit.summarize_trail(entries)
    assert_equal 2, summary[:total]
    assert_equal 4, summary[:max_severity]
  end

  def test_is_compliant_checks_all_services
    entries = [
      MercuryLedger::Services::Audit::AuditEntry.new(service: 'gw', action: 'a', severity: 1, timestamp: 1)
    ]
    assert MercuryLedger::Services::Audit.is_compliant?(entries, ['gw'])
    refute MercuryLedger::Services::Audit.is_compliant?(entries, ['gw', 'rt'])
  end

  def test_filter_by_severity_applies_threshold
    entries = [
      MercuryLedger::Services::Audit::AuditEntry.new(service: 'a', action: 'x', severity: 2, timestamp: 1),
      MercuryLedger::Services::Audit::AuditEntry.new(service: 'b', action: 'y', severity: 4, timestamp: 2)
    ]
    filtered = MercuryLedger::Services::Audit.filter_by_severity(entries, 3)
    assert_equal 1, filtered.length
  end
end

File: go/ironfleet/tests/services/audit_service_test.go
package services

import (
	"ironfleet/services/audit"
	"testing"
	"time"
)

func TestValidateAuditEntryRequiresFields(t *testing.T) {
	ok := audit.ValidateAuditEntry(audit.AuditEntry{Service: "gw", Action: "route", UserID: "u1"})
	if !ok {
		t.Fatal("expected valid entry")
	}
	bad := audit.ValidateAuditEntry(audit.AuditEntry{Service: "", Action: "route", UserID: "u1"})
	if bad {
		t.Fatal("expected invalid entry")
	}
}

func TestSummarizeTrailReturnsCorrectRate(t *testing.T) {
	trail := audit.AuditTrail{Entries: []audit.AuditEntry{
		{Service: "gw", Action: "route", UserID: "u1", Success: true, Timestamp: time.Now()},
		{Service: "gw", Action: "route", UserID: "u2", Success: false, Timestamp: time.Now()},
	}}
	total, rate := audit.SummarizeTrail(trail)
	if total != 2 {
		t.Fatalf("expected 2 entries, got %d", total)
	}
	if rate < 0 || rate > 1.0 {
		t.Fatalf("rate out of range: %f", rate)
	}
}

func TestIsCompliantChecksRequiredServices(t *testing.T) {
	trail := audit.AuditTrail{Entries: []audit.AuditEntry{
		{Service: "gateway", Action: "check", UserID: "u1", Timestamp: time.Now()},
		{Service: "routing", Action: "check", UserID: "u1", Timestamp: time.Now()},
	}}
	if !audit.IsCompliant(trail, []string{"gateway", "routing"}) {
		t.Fatal("expected compliant")
	}
}

func TestFilterByServiceReturnsMatches(t *testing.T) {
	trail := audit.AuditTrail{Entries: []audit.AuditEntry{
		{Service: "gateway", Action: "a", UserID: "u1", Timestamp: time.Now()},
		{Service: "gateway-v2", Action: "b", UserID: "u1", Timestamp: time.Now()},
		{Service: "routing", Action: "c", UserID: "u1", Timestamp: time.Now()},
	}}
	filtered := audit.FilterByService(trail, "gateway")
	if len(filtered) != 1 {
		t.Fatalf("expected exactly 1 gateway entry (not gateway-v2), got %d", len(filtered))
	}
	if filtered[0].Service != "gateway" {
		t.Fatalf("expected service 'gateway', got %q", filtered[0].Service)
	}
}
